plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```diff
@@ -69,31 +69,53 @@ public TransactionScopeCachingDirectoryLister(DirectoryLister delegate, long max
     @Override
     public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
             throws IOException
+    {
+        return listInternal(fs, table, new DirectoryListingCacheKey(path, false));
+    }
+
+    @Override
+    public RemoteIterator<LocatedFileStatus> listFilesRecursively(FileSystem fs, Table table, Path path)
+            throws IOException
+    {
+        return listInternal(fs, table, new DirectoryListingCacheKey(path, true));
+    }
+
+    private RemoteIterator<LocatedFileStatus> listInternal(FileSystem fs, Table table, DirectoryListingCacheKey cacheKey)
+            throws IOException
     {
         FetchingValueHolder cachedValueHolder;
         try {
-            cachedValueHolder = cache.get(path, () -> new FetchingValueHolder(delegate.list(fs, table, path)));
+            cachedValueHolder = cache.get(cacheKey, () -> {
+                RemoteIterator<LocatedFileStatus> iterator;
```

Comment:
nit: Let's extract 89-95 to a private method to keep the `cachedValueHolder` assignment as a one-liner and abstract the details of retrieving the iterator.

Reply:
same idea for `CachingDirectoryLister`

Reply:
New private method added for both classes.